SSA/Core/MLIRSyntax/Parser.lean

```diff
@@ -0,0 +1,132 @@
+  https://leanprover-community.github.io/mathlib4_docs/Std/Util/TermUnsafe.html#Std.TermUnsafe.termUnsafe_
+ -/
+def elabOp (env : Lean.Environment) (stx : Lean.Syntax) : EIO ParseError (MLIR.AST.Op 0) :=
+  unsafe elabIntoEIO (α := MLIR.AST.Op 0) env op0Expr stx
+
+def mkParseFun (syntaxcat : Name)
+    (ntparser : Syntax → EIO ParseError ParseOutput)
+    (s : String) (env : Environment) : EIO ParseError ParseOutput := do
+  match Parser.runParserCategory env syntaxcat s with
+    | .error msg => throw msg
+    | .ok syn => ntparser syn
+
+private def mkNonTerminalParser (syntaxcat : Name)
+    (ntparser : Syntax → EIO ParseError ParseOutput)
+    (env : Environment) (s : String) : EIO ParseError ParseOutput :=
+  let parseFun := mkParseFun syntaxcat ntparser
+  parseFun s env
+
+-- TODO: do we need this copy of the environment, or can I remove it from one of the two?
+def regionParser : @ParseFun (MLIR.AST.Region 0) :=
+  fun env : Lean.Environment => mkNonTerminalParser `mlir_region (elabRegion env) env
+
+private def parseFile (env: Lean.Environment)
+    (parser: @ParseFun ParseOutput)
+    (filepath: System.FilePath) : IO (Option ParseOutput) := do
+  let lines ← IO.FS.lines filepath
+  let fileStr := "\n".intercalate lines.toList
+  let parsed ← EIO.toIO' <| parser env fileStr
+  match parsed with
+  | .ok parseOutput => return parseOutput
+  | .error msg => IO.println s!"Error parsing {filepath}:\n{msg}"; return none
+
+private def isFile (p: System.FilePath) : IO Bool := do
+  return (← p.metadata).type == IO.FS.FileType.file
+
+def runParser (parser : @ParseFun ParseOutput) (fileName : String) : IO (Option ParseOutput) := do
+  /- We expect our parser to run with `lake exec`, which sets `LEAN_PATH` to be a colon separated
+    list of package paths.
+    We parse the package paths and set the search path accordingly.
+    For more, see `lake env help.` -/
+  let packagePaths : List String :=
+    match (← IO.getEnv "LEAN_PATH") with
+    | .none => []
+    | .some colonSeparatedPaths => colonSeparatedPaths.splitOn ":"
+  if packagePaths.isEmpty then
+    throw <| IO.userError "Expected `LEAN_PATH` environment variable to be set. Are you running via `lake exec opt`?"
+  initSearchPath (← Lean.findSysroot) packagePaths
+  let modules : Array Import := #[⟨`SSA.Core.MLIRSyntax.EDSL, false⟩]
+  let env ← importModules modules {}
+  let filePath := System.mkFilePath [fileName]
+  if !(← isFile filePath) then
+    throw <| IO.userError s!"File {fileName} does not exist"
+  parseFile env parser filePath
+
+def parseRegionFromFile (fileName : String)
+    (regionParseFun : MLIR.AST.Region 0 → Except ParseError α) : IO (Option α)  := do
+  let ast ← runParser regionParser fileName
+  match Option.map regionParseFun ast with
+  | .some (.ok res) => return res
+  | .some (.error msg) => throw <| IO.userError s!"Error parsing {fileName}:\n{msg}"; return none
+  | .none => return none
```

Comment:
My feeling is that the contents of this file as well can be moved into `SSA/Projects/InstCombine/LLVM/Opt.lean`, and can be judiciously refactored out when necessary.

Reply:
Hmm, I disagree. This is not about the verified `opt` executable, this is a library about parsing. I wouldn't mind it if this had been done all in the first place and could be refactored later, but refactoring so that we can later refactor again seems silly :)